docs/adv_settings.md
# Advanced (hidden) settings

There are some settings not in the CLI or config file template that can be set in your config (and/or done with `--override`).

- `dbcache_dir = None`: Where to store the file database. Default is `<rclone cache dir>/DFB/<_uuid from config>.db`
- `disable_refresh = False`: Do not allow refresh

docs/compare_settings.md
# Compare Setting Guidance

## Default

Keep the defaults of `"auto"` for most. They are

    'compare'
    'dst_compare'
    'renames'
    'dst_renames'
    'get_modtime'
    'get_hashes'

Note that the auto is using "mtime" if and when possible for compare. If "mtime" is not possible, disable renames

## Comparison and Rename Attributes

The attributes for comparison and for renames are user settable. If both remotes support hashes, it almost always best to use them. And if the source is slow to list ModTime, you can also set `get_modtime = False`. If remotes support ModTime and it is fast, that is a decent choice for both compare and renames.

Generally speaking, comparisons and renames are actually source-to-source because the source values are saved. However, if run with `--refresh`, then comparisons and move-tracking are source-to-dest. In that case, you can set `dst_compare` and `dst_renames`. 

<table>
    <tr>
        <th>Source</th>
        <th>Destination</th>
        <th><code>compare</code></th>
        <th><code>dst_compare</code></th>
        <th><code>renames</code></th>
        <th><code>dst_renames</code></th>
        <th>Comment</th>
    </tr>
    <tr>
        <td>Local</td>
        <td>
            Any remote that supports ModTime including 
            local, B2, OneDrive, DropBox, [Google] Drive,
            [S]FTP, <strong>certain</strong> WebDAV. <strong>NOT</strong> S3.
        </td>
        <td>'mtime'</td>
        <td>None</td>
        <td>'mtime'</td>
        <td>None</td>
        <td>Use 'mtime' since it easy and fairly reliable</td>
    </tr>
        <td>Local</td>
        <td>
            S3, <em>Regular</em> WebDAV
        </td>
        <td>'mtime'</td>
        <td>'size'</td>
        <td>'mtime'</td>
        <td>False</td>
        <td>
            Use 'mtime' when using past source data but switch to size
            when using <code>dst_</code> since mtime is either super slow
            (S3) or unreliable (WebDAV). Since 'size' is a poor file tracker,
            disable renames in that case.
        </td>
    </tr>
    <tr>
        <td colspan="2">
            <strong>Identical</strong> cloud-to-cloud that support fast hashing (i.e. not SFTP). Or if the same hash is supported (e.g. S3 to Azure)
        </td>
        <td>'hash'</td>
        <td>None</td>
        <td>'hash'</td>
        <td>None</td>
        <td>Use 'hash' since it is fast and reliable</td>
    </tr>
    <tr>
        <td colspan="2">
            <strong>Different</strong> cloud-to-cloud that support fast hashing (i.e. not SFTP)
        </td>
        <td>'hash'</td>
        <td>'size'</td>
        <td>'hash'</td>
        <td>None</td>
        <td>
            Won't have hashes on both. Can change 'size' to 'mtime' if both 
            support it and one is not S3 (where 'mtime' is slow)
        </td>
    </tr>
    <tr>
        <td>Cloud that supports fast hashing</td>
        <td>Local or SFTP</td>
        <td>'hash'</td>
        <td>'size'</td>
        <td>'hash'</td>
        <td>None</td>
        <td>
            Can change 'size' to 'mtime' if  
            not S3 (where 'mtime' is slow)
        </td>
    </tr>
    <tr>
        <td>regular WebDAV, Mega, Seafile, any other non-mod-time remote</td>
        <td>Anything</td>
        <td>'size'</td>
        <td>None</td>
        <td>False</td>
        <td>False</td>
        <td>
            Only use size to detect changed files. Good enough most of the time but unreliable for move detection
        </td>
    </tr>
</table>

**NOTE**: Even if compares are *not* `'mtime'`, all restores will do `mtime` if possible!

## Alternative S3 source settings

If your source is S3 and you do not want to work off of "hash" for whatever reason (some non-rclone tools do not set it always) and don't want to fall back to 'size' or the slow 'mtime', one option is to add the flag `--use-server-modtime` and then `'mtime'` for the compare. This will detect changes just fine!


changelog.md
# Changelog

(newest on top)

## 20240106.0

- Added documentation that dfb-mount needs libfuse2 and will not work on libfuse3.
    ```
    $ sudo apt-get update -y
    $ sudo apt-get install -y libfuse2
    ```
- Added tool called `dfb-link` that will symlink the correct files. This can be used instead of dfb-mount if you do not want to use FUSE.
- Minor code changes to accommodate using the timestamp tool outside of dfb.

## 20231227.0

- `refresh` is now its own command and has been removed from all other commands except for `backup`
- Added better help for setting up attributes to readme

## 20230705.0

- Do not write debug to file except in testing. This was a leftover artifact that mostly didn't matter except when pruning a large backup, it was writing a line for 2-3x the number of files. 
- Better pypy support:
    - Added a dir creation that *shouldn't* be needed but seems to be under pypy. I will try to figure that out
    - Testing fails due to "open files". I think pypy should work but it is only lightly tested.

## 20230703.0

- Adds a `-N N, --keep-versions N` flag to prune. This lets you control versions more carefully and do things like:
    - Prune more than 15 copies: `prune now -N 15`
- Adds "`now`" as a valid timestamp.
- Adds the `ID:` to the log. Useful if messing with caches
- Adds "`total_size`" to `ls -ll`

## 20230630.0

- Removes some features that are no longer necessary with rclone 1.63:
    - Removes `dst_atomic_transfer` and supporting workaround. If you use a remote that is not atomic (e.g. sftp, ftp, local) and you do *not* want rclone to handle this, add `--inplace` to flags
    - see note below
- Adds a minimum rclone version and checks
- **Minimumum rclone version is 1.63**

Future versions will remove the special symlink handling but there are some bugs on the rclone side to be worked out first

## 20230623.0

- Handles the situation when using `compare = 'hash'` but the checksum/hash is still missing, as can happen with some remotes. Falls back to `'size'` compare similar to rclone's behavior.
    - Settable with `error_on_missing_hash` config
- More robust hash database even around missing hashes (unlikely to actually happen for the kind of remotes you would want to use a database but this should handle it.)
- Fixed bug with `reuse_hashes='size'` still caring about `mtime`

## 20230606.0

- Added the option to track moves and use a [server-side] copy instead of a reference. This only makes sense when the remote support server-side copy (not verified by dfb). It enables a cleaner representation without using bandwidth
- Tests for new capability

## 20230508.0

- Fixed removed feature.

## 20230507.0

- Adds the `disable_prune` (default to False) configuration option so that it is harder to accidentally prune. Makes it effectively append-only unless very, very explicit.
- Minor documentation updates

## 20230502.0

- Can specify `--deleted` (or `--del`) twice for `ls` and `snaphot` to *only* include deleted items

## 20230501.0

- Adds version count to `ls -ll`

## 20230429.0

- Adds the option to upload snapshot files with each run. These are functionally the same as calling `snapshot --only <timestamp> --deleted`
- Added `--deleted` to snapshot
- Bug Fixes
    - source and dest labels mixed up in restore logs

## 20230411.0 (BETA)

- Bug Fix

## 20230407.0 (BETA)

- If any part fails, it'll keep going until deletes and references happen
- Adds pruning only a subdir. Still have to compute all prunes to ensure there aren't references but will then filter them.

## 20230402.0 (BETA)

- Adds support for symlinks on a local source. It is not super robust but it works for most cases. Note that dfb does not restore the symlinks as symlinks. It restores them as `.rclonelink` files which can be easily turned into symlinks later. See an example in the readme

## 20230331.0 (BETA)

- **EXPERIMENTAL** FUSE mount that overlays an rclone mount (does not serve the remote directly)
- Cleanup, documentation, etc
- Fixed version numbering to validate PEP440 and other build errors.
    - This is still beta, but just isn't in the date-based version number

## 20230330.0.BETA

- Changed how references are written and saved. The change will not break existing references but older references will not work in the upcoming mounting tools
- Minor other updates

## 20230327.0.BETA


dfb/configuration.py
import sys, os
import time
import tempfile
import uuid
import copy
import logging
import io
import hashlib, base64
from functools import partial, partialmethod
from pathlib import Path
from threading import Lock

logger = logging.getLogger(__name__)

_r = repr

LOCK = Lock()
_TEMPDIR = False  # Just used in testing

INF = float("inf")


class NotDFBFilter(logging.Filter):
    def __init__(self, include_serve=False):
        self.include_serve = include_serve
        super().__init__()

    def filter(self, record):
        if record.name.endswith("-rc-server") and not self.include_serve:
            return False
        return record.name.startswith("dfb.")


def init_logging(logfile, debuglogfile, verbosity):
    """
    Start logging. If _TEMPDIR is set, create a second one that always saves with
    more detail
    """
    USE_DEBUGFILE = bool(_TEMPDIR)

    levels = [logging.WARN, logging.INFO, logging.DEBUG]

    verbosity = min([len(levels), max([0, verbosity])])  # +1 for rc server
    if verbosity == len(levels):
        not_dfb_filter = NotDFBFilter(include_serve=True)
        verbosity -= 1
    else:
        not_dfb_filter = NotDFBFilter()

    level = levels[verbosity]

    formatter = logging.Formatter(
        fmt="%(asctime)s:%(levelname)s: %(message)s",
        datefmt="%Y-%m-%d %H:%M:%S",  # "%z" removed
    )

    # Set up handlers with the level since the root_logger *may* be set lower
    file_handler = logging.FileHandler(logfile)
    file_handler.setFormatter(formatter)
    file_handler.setLevel(level)
    file_handler.addFilter(not_dfb_filter)

    stream_handler = logging.StreamHandler(stream=sys.stderr)
    stream_handler.setFormatter(formatter)
    stream_handler.setLevel(level)
    stream_handler.addFilter(not_dfb_filter)

    root_logger = logging.getLogger()
    root_logger.handlers.clear()  # Need to clear them  for testing

    root_logger.setLevel(level=0 if USE_DEBUGFILE else level)
    root_logger.addHandler(file_handler)
    root_logger.addHandler(stream_handler)

    if USE_DEBUGFILE:
        dformatter = logging.Formatter(
            fmt=(
                "%(asctime)s.%(msecs)d:"  # https://stackoverflow.com/a/7517430/3633154
                "%(name)s:"
                # "%(module)s:"
                "%(lineno)d:"
                "%(funcName)s:"
                "%(levelname)s: "
                "%(message)s"
            ),
            datefmt="%Y%m%d%H%M%S",
        )

        dfile_handler = logging.FileHandler(debuglogfile)
        dfile_handler.setFormatter(dformatter)
        dfile_handler.setLevel(logging.DEBUG)
        dfile_handler.addFilter(NotDFBFilter(include_serve=True))

        root_logger.addHandler(dfile_handler)


def clean_config_id(config_id):
    """Cleans it up to alpha-numeric only"""
    config_id0 = config_id
    allowed = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_-[]"
    config_id = "".join(c if c in allowed else "=" for c in config_id)

    if len(config_id) > 40:
        md5 = base64.urlsafe_b64encode(
            hashlib.md5(config_id.encode()).digest()
        ).decode()
        config_id = f"{config_id[:20]}.{md5[:8]}.{config_id[-20:]}"

    if config_id0 != config_id:
        logger.debug(f"config_id changed from {_r(config_id0)} to {_r(config_id)}")

    return config_id


class ConfigError(ValueError):
    pass


class Config:
    def __init__(self, configpath, tmpdir=None, verbosity=1, add_params=None):
        from . import nowfun, __version__, __git_version__

        self._config = {"_configpath": configpath, "verbosity": verbosity}
        try:
            self.configpath = Path(configpath).resolve()  # make it absolute
        except FileNotFoundError:
            raise FileNotFoundError(f"Couldn't find {_r(configpath)}")
        self.add_params = add_params or {}

        self.now = nowfun()

        if _TEMPDIR:  # Testing
            self.tmpdir = Path(_TEMPDIR)
        elif not tmpdir:
            self.tmpdir = Path(tempfile.TemporaryDirectory().name)
        else:
            self.tmpdir = Path(tmpdir) / f"{int(self.now.dt)}"

        self.tmpdir.mkdir(parents=True, exist_ok=True)

        # Start the logging
        self.logfile = self.tmpdir / "log.log"
        self.debuglogfile = self.tmpdir / "debug.log"
        init_logging(self.logfile, self.debuglogfile, verbosity)

        logger.info(f"DFB ({__version__})")
        if __git_version__:
            logger.info(f" {__git_version__['version']} {__git_version__['origin']}")
        logger.info(f"Now: {self.now.obj.astimezone().isoformat()}")
        logger.info(f"Backup Timestamp: {self.now.dt}Z")
        logger.info(f"config path: '{self.configpath}'")
        logger.info(f"tmpdir: {str(self.tmpdir)}")

    def _write_template(self, force=False):
        from . import __version__

        txt = TEMPLATE
        txt = txt.replace("__VERSION__", __version__)
        txt = txt.replace("__UUID4__", str(uuid.uuid4()))

        if self.configpath.exists() and not force:
            raise ValueError(
                f"Path '{self.configpath}' exists. "
                "Specify a different path, move the existing file "
                "or use --force-overwrite"
            )
        self.configpath.parent.mkdir(parents=True, exist_ok=True)
        self.configpath.write_text(txt)

        os.chmod(self.configpath, self.configpath.stat().st_mode | 0o100)

        logger.debug(f"Wrote template config to {self.configpath}")

    def parse(self, override_txt=""):
        from .rclonerc import RC

        if self.configpath is None:
            raise ValueError("Must have a config path")

        # Passed to the config file
        self._config["stats"] = 30  # Fixed. Undocumented that this can be set
        self._config["os"] = os
        self._config["Path"] = Path
        self._config["log"] = lambda x: logger.info(f"config: {x}")
        self._config["print"] = lambda x: logger.info(f"config: {x}")
        self._config["__file__"] = self.configpath
        self._config["__dir__"] = self.configpath.parent
        self._config["DELENV"] = RC.DELENV
        self._config["clean_config_id"] = clean_config_id

        self._config.update(self.add_params)

        self._hidden_keys = set(self._config)  # to be removed in repr
        self._hidden_keys.difference_update({"configpath", "verbosity"})

        junk = {}
        exec("", junk)

        exec(TEMPLATE, self._config)  # Defaults
        self._config_keys = [
            k
            for k in self._config
            if (k not in junk and k not in self._hidden_keys and not k.startswith("_"))
        ]
        self._config_keys.append("destpaths")

        # Read then change dir
        config_txt = self.configpath.read_text()
        # os.chdir(self._config["__dir__"])

        # Set the override_txt before AND after so that you can set other things
        cfg = [
            "pre,post = True,False",
            override_txt,
            config_txt,
            "pre,post = False,True",
            override_txt,
        ]
        exec("\n".join(cfg), self._config)

        for key in junk:
            self._config.pop(key, 0)

        self._validate()

        logger.info(f"ID: {self.config_id}")
        logger.debug(f"Read config {_r(str(self.configpath))}")
        for k in self._config_keys:
            if k not in self._config:
                continue
            dispval = self._config[k]
            if k == "rclone_env":
                dispval = {
                    n: (k if n != "RCLONE_CONFIG_PASS" else "**REDACTED**")
                    for n, k in dispval.items()
                }
            logger.debug(f"   {k} = {_r(dispval)}")

        # Set these up because all methods will use them
        settings = dict(
            rclone_exe=self.rclone_exe,
            universal_flags=self.rclone_flags,
            universal_env=self.rclone_env,
            # Even if using something like Google Drive which allows duplicates,
            # we won't *ever* be overwriting files so this is okay.
            no_check_dest=True,
        )
        if self.metadata:
            settings["universal_flags"].append("--metadata")
            # self.rclone_flags.append("--metadata")

        # Set up rclone objects. For the most part, we use the RC interface
        # but there are a few things where the CLI is better or easier. These may
        # go away in future versions. The CLI interface (rclonecli.py) has many
        # "features" that are eclipsed by rc, but a few are better on CLI.
        #
        # Add these here since we will need them throughout the eval. Note that
        # rc is using the rclone server while (src/dst)_rclone is making CLI
        # calls to rclone.
        from .rclonecli import RcloneCLI
        from .rclonerc import RC

        self.rc = RC(
            rclone_exe=self.rclone_exe,
            serve_flags=self.rclone_flags
            + ["-vv"],  # always include verbose but filter later
            rclone_env=self.rclone_env,
        )

        self.src_rclone = RcloneCLI(self._config["src"], **settings)
        self.dst_rclone = RcloneCLI(self._config["dst"], **settings)
        # Monkey patch the debug
        self.src_rclone.debug = lambda x: logger.debug(f"src-rclone: {x}")
        self.dst_rclone.debug = lambda x: logger.debug(f"dst-rclone: {x}")

        return self

    def _validate(self):
        """
        Validate config
        """
        from .rclonerc import FILTER_FLAGS

        if self.src == "<<MUST SPECIFY>>":
            raise ConfigError("Must specify 'src'")
        if self.dst == "<<MUST SPECIFY>>":
            raise ConfigError("Must specify 'dst'")

        allowed = {
            "compare": {"mtime", "size", "hash", "auto"},
            "dst_compare": {"mtime", "size", "hash", "auto", None},
            "renames": {"size", "mtime", "hash", "auto", False, None},
            "dst_renames": {"size", "mtime", "hash", "auto", False, None},
            "links": {"skip", "link", "copy"},
            "rename_method": {"reference", "copy", False, None},
            "get_modtime": {True, False, "auto"},
            "get_hashes": {True, False, "auto"},
        }

        for key, values in allowed.items():
            val = self._config[key]
            if val not in values:
                msg = f"Allowed values for '{key}' are {values}. Specified {repr(val)}"
                raise ConfigError(msg)
        ff = FILTER_FLAGS.union(["--one-file-system"])
        if badflags := ff.intersection(self.rclone_flags):
            msg = f"May not have {badflags} in 'rclone_flags'. Use 'filter_flags'"
            raise ConfigError(msg)

        # These will also set to auto if the compare or renames is auto
        self._config["dst_compare"] = (
            self._config["dst_compare"] or self._config["compare"]
        )
        if self._config["dst_renames"] is None:  # explicit because could be False
            self._config["dst_renames"] = self._config["renames"]

        if self._config["links"] == "copy":
            self._config["rclone_flags"].append("--copy-links")

        if not self._config.get("dst_atomic_transfer", True):
            logger.info(
                "WARNING: 'dst_atomic_transfer = False' is deprecated since rclone 1.63 "
                "handles it for non-atomic remotes"
            )

        # Set the config_id but give preference to _uuid
        self._config["config_id"] = clean_config_id(
            self._config.get("_uuid", self._config["config_id"])
        )

        if mrs := self._config["min_rename_size"]:
            from .utils import parse_bytes

            self._config["min_rename_size"] = mrs1 = parse_bytes(mrs)
            logger.debug(f"Parsed min_rename_size {_r(mrs)} as {_r(mrs1)} bytes")

    def _set_auto(self):
        sf = self.rc.features(self.src)
        df = self.rc.features(self.dst)

        src_mtime = sf["Precision"] < 1.1e9 and not sf["Features"]["SlowModTime"]
        dst_mtime = df["Precision"] < 1.1e9 and not df["Features"]["SlowModTime"]

        if self.compare == "auto":  # src to src
            self.compare = "mtime" if src_mtime else "size"
            logger.debug(f"auto-setting 'compare' to {_r(self.compare)}")

        if self.dst_compare == "auto":  # src to dst
            # don't *just* do self.compare since it could be hash
            if self.compare != "size" and src_mtime and dst_mtime:
                self.dst_compare = "mtime"
            else:
                self.dst_compare = "size"

            logger.debug(f"auto-setting 'dst_compare' to {_r(self.dst_compare)}")

        if self.renames == "auto":  # src to src
            self.renames = "mtime" if src_mtime else False
            logger.debug(f"auto-setting 'renames' to {_r(self.renames)}")

        if self.dst_renames == "auto":  # src to dst
            # don't *just* do self.renames since it could be hash or False
            if self.renames != "size" and self.renames and src_mtime and dst_mtime:
                self.dst_renames = "mtime"
            else:
                self.dst_renames = False

            logger.debug(f"auto-setting 'dst_renames' to {_r(self.dst_renames)}")

        if self.get_modtime == "auto":
            self.get_modtime = src_mtime
            logger.debug(f"auto-setting 'get_modtime' to {_r(self.get_modtime)}")

        if self.get_hashes == "auto":
            # self.get_hashes = sf["Hashes"] and not sf["Features"]["SlowHash"]
            # if sf['String'].startswith('S3 '):
            #    logger.warning(
            #        "src is S3 remote. S3 doesn't *always* provide hashes without "
            #        "additional API calls. Setting get_hashes to False to be safe"
            #    )
            #    self.get_hashes = False
            # logger.debug(f"auto-setting 'get_hashes' to {_r(self.get_hashes)}")

            # to be safe, making this always false. Users should set this if they want it
            self.get_hashes = False
            logger.debug(f"setting 'get_hashes' to False regardless of remotes")

    def __getattr__(self, attr):
        return self._config[attr]

    def __setattr__(self, attr, value):
        if attr.startswith("_"):
            return super(Config, self).__setattr__(attr, value)

        self._config[attr] = value

    def __repr__(self):
        # Need to watch out for RCLONE_CONFIG_PASS in rclone_env
        # make a copy of the dict fixing that one but do not
        # just do a deepcopy in case the user imported modules
        cfg = copy.copy(self._config)
        cfg["rclone_env"] = cfg["rclone_env"].copy()

        if "RCLONE_CONFIG_PASS" in cfg["rclone_env"]:
            cfg["rclone_env"]["RCLONE_CONFIG_PASS"] = "**REDACTED**"

        contents = ", ".join(
            f"{k}={_r(cfg[k])}" for k in self._config_keys if k in self._config
        )
        return f"Config({contents})"


TEMPLATE = r'''#!/usr/bin/env dfbshebanged
"""
DFB Config File

This configuration file is read as Python so things can be customized as
desired. With few exceptions, any missing items will go to the defaults
already specified.

rclone flags should always be a list. 
Example: `--exclude myfile` will be ['--exclude','myfile']

Defaults are sensible for a local source. Change as needed for others.

A few modules, including `os` and `Path = pathlib.Path` are already loaded along
with `logger.info()` and `logger.debug()`

Also defines:
    __file__ : Absolute path of the config file. pathlib.Path
    __dir__ : Absolute path of the config file parent. pathlib.Path
    subdir : Value of '--subdir' if specified, otherwise an empty string.

ALL LOCAL PATHS SHOULD BE ABSOLUTE
"""
##############################################
##              Basic Settings              ##
##                (required)                ##
##############################################

# Specify the source and destination rclone remotes
src = "<<MUST SPECIFY>>"
dst = "<<MUST SPECIFY>>"

##############################################
##                 Filters                  ##
##############################################

# Specify flags used for filtering. See https://rclone.org/filtering/
#
# If using flags like '--filter-from', they should be absolute paths. Example:
#   ["--filter-from", __dir__ / "myfilters.txt"]
#
# Note that when backups use --subdir, the paths specified here may be incorrect to the
# source. The variable 'subdir' is defined to assist. USE WITH CAUTION
filter_flags = []


##############################################
## Comparison and Rename-Tracking Settings  ##
##           (defaults to "auto")           ##
##############################################

# src-to-src comparison to determine changes. Auto tries to use mtime if supported.
compare = "auto"  # "size", "mtime", "hash", "auto"

# src-to-dst comparison such as after a refresh without snapshots. None uses 'compare'
dst_compare = None  # "size", "mtime", "hash", "auto", None

# Rename tracking src-to-src. False disables it. "size" is risky!
renames = "auto"  # "size", "mtime", "hash", "auto", False

# Rename tracking src-to-dst. None uses 'renames'
dst_renames = None  # "size", "mtime", "hash", "auto", False, None

# Rename files with server-side-copy or with reference file
rename_method = "reference"  # 'reference', 'copy'

# If files are less than this size, do not track the rename. Can specify with
# normal prefixes or an integer byte count (e.g. 2097152, "2 KiB", "10 MB", "15 MiB")
min_rename_size = 0

##############################################
##             rclone Settings              ##
##         (optional, intermediate)         ##
##############################################

# Flags and environment for general rclone usage.
# Examples include --config (or RCLONE_CONFIG).
# Example for config password
#   > from getpass import getpass
#   > rclone_env = {"RCLONE_CONFIG_PASS": getpass("Password: ")}
# Not all flags are compatible. Do not use --links (see below)
rclone_flags = []
rclone_env = {}

# Flags for refresh specifically. Example: --fast-list
dst_list_rclone_flags = []

# Executable
rclone_exe = "rclone"

##############################################
##          Intermediate Settings           ##
##                (optional)                ##
##############################################

# How to handle links on local src. 'link' makes .rcloneline files and
# 'copy' copies the referent. See https://github.com/rclone/rclone/issues/6855
links = "link"  # {'skip','link','copy'}

# Number of transfers. Use remote-specific flags for additional control such as
# --s3-upload-concurrency.
concurrency = os.cpu_count()

# Tolerance on mtimes
dt = 1.0  # seconds

# Whether to always get src mtime. Can be expensive and slow on some remotes.
get_modtime = "auto"  # True, False, "auto"

# Allow missing hash. Falls back to "size"
error_on_missing_hash = False

# String or list of strings of hash types. None uses all available on a remote
hash_type = None

# Whether to always request hashes. "auto" maps to False for now
get_hashes = False  # True, False, "auto"

# Request and transfer (if possible) metadata. Metadata is also stored in the snapshot
# files in case the dst doesn't support it. However, it must be restored manually if not
# supported. Changes in file metadata will *not* force a backup again.
metadata = True

# Specify additional log destinations. These should be full rclone remote directories
# to save the log in addition to the remote. String or list of strings
log_dest = None

# Unique name for this config file. Default is just based on the src and dst. Will be
# processed and cleaned as needed
config_id = f"{src}-{dst}"

# Specify where to store the file database. Default (None) is
# `<rclone cache dir>/DFB/<config_id>.db`.
dbcache_dir = None

##############################################
##             Disable Features             ##
##     Use --override to undo. Example:     ##
##    --override "disable_prune = False"    ##
##############################################

# Pruning is the only distructive process in dfb. In order to make sure it isn't done by
# accident, setting this to True will force all pruneing to be as if called with
disable_prune = False

# While not destructive, refresh can be very slow on some remotes, especially with many
# references and/or not using snapshots
disable_refresh = False

##############################################
##       Pre- and Post-Shell Options        ##
##                (optional)                ##
##############################################

## Pre-, Post- and fail- shell commands to run
# Specify shell code to be evaluated before and/or after running dfb. Note
# these are all run from the current directory but $CONFIGDIR environment will be set.
# STDOUT and STDERR will be captured. Note that there is no validation or
# security of the inputs. These are not actually called if using dry-run.
#
# The post shell call also has the "$STATS" environment variable defined which has
# the run statistics including timing (which will be different than the final since
# the logs will not yet have been dumped)
#
# Can be specified as the following:
#
#     string : Run with shell=True Can cd as needed (including `cd $CONFIGDIR`) and
#              can be multiple lines and multiple commands.
#     list   : Will execute with shell=False in the current directory
#     dict   : Specify subprocess.Popen flags plus the keyword 'cmd'. YOU decide if
#              shell should be True or False based on 'cmd'. dfb will override settings
#              for std(out/err). Will update os.environ with any 'env' settings.
#
# If the final cmd if specified directly as a list or a list inside the dict, each
# command will be run through C-style formatting of the environ. C-Style is used to not
# break the more-common str.format (or f-string) formats that may exist.
# os.environ may be updated with `env` inside of a dict.
#
# Example:
#   post_shell = 'echo "$STATS"'
#   post_shell = ["echo","%(STATS)s"]
# will look the same
pre_shell = ""
post_shell = ""

# Specify whether or not to allow an error in the shell commands above to continue
stop_on_shell_error = False

# If, and only if, the run fails, this will get called. It will not be logged or
# have any information about the failure however the temporary log, $LOGPATH, and
# debug log, $DEBUGPATH, could be read if interested. Useful for sending alert of
# failed run. Same rules as above about being specified as a list
fail_shell = ""


#######
# This should only be changed by the user when migrating from an older config
# to a newer one. Just because the current version of dbf and the version
# below do not match, it does not mean the run won't work.
_version = "__VERSION__"
'''
